File: bestia-messages/src/main/java/net/bestia/messages/MessageTypeIdResolver.java
package net.bestia.messages;

import com.fasterxml.jackson.annotation.JsonTypeInfo.Id;
import com.fasterxml.jackson.databind.DatabindContext;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.jsontype.impl.TypeIdResolverBase;
import com.fasterxml.jackson.databind.type.TypeFactory;
import org.reflections.Reflections;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.util.HashMap;
import java.util.Set;

/**
 * Custom TypeId Resolver for message objects. Upon start it looks for all
 * messages which inherit from Message and determine their id. It stores it and
 * uses this serialize the messages.
 * 
 * @author Thomas Felix
 *
 */
public class MessageTypeIdResolver extends TypeIdResolverBase {

	private final static Logger log = LoggerFactory.getLogger(MessageTypeIdResolver.class);

	private final TypeFactory typeFactory = TypeFactory.defaultInstance();
	private static final HashMap<String, Class<? extends MessageId>> idToClass = new HashMap<>();
	private static final HashMap<Class<? extends MessageId>, String> classToId = new HashMap<>();

	static {

		// Find all classes implementing the message interface.
		final Reflections reflections = new Reflections("net.bestia.messages");
		final Set<Class<? extends MessageId>> messages = reflections.getSubTypesOf(MessageId.class);

		// Instantiate the message classes to get their message id from the
		// method and store it for later serialization and deserialization.
		for (Class<? extends MessageId> msg : messages) {

			// Avoid abstract classes.
			if (Modifier.isAbstract(msg.getModifiers())) {
				continue;
			}

			try {
			  final Field[] fields = msg.getFields();
			  log.debug(fields.toString());
				final Field messageId = msg.getDeclaredField("MESSAGE_ID");
				final String key = (String) messageId.get(null);

				idToClass.put(key, msg);
				classToId.put(msg, key);

				log.trace("Found Message.class: {} - {}", key, msg.toString());

			} catch (Exception e) {
				log.error("Could not get static MESSAGE_ID field from Message class: "+ msg.getName() +". Serialization and deserialization will fail.", e);
				System.exit(1);
			}
		}
	}

	private JavaType baseType;

	/**
	 * Finds all IDs of the messages and registers them for later
	 * identification.
	 */
	@Override
	public void init(JavaType bt) {
		super.init(bt);
		baseType = bt;

	}

	@Override
	public Id getMechanism() {
		return Id.CUSTOM;
	}

	@Override
	public String idFromValue(Object value) {
		return idFromValueAndType(value, value.getClass());
	}

	@Override
	public String idFromValueAndType(Object value, Class<?> suggestedType) {
		return classToId.get(suggestedType);
	}

	@Override
	public JavaType typeFromId(DatabindContext context, String id) {
		final Class<? extends MessageId> clazz = idToClass.get(id);
		return typeFactory.constructSpecializedType(baseType, clazz);
	}
}

File: bestia-messages/src/main/kotlin/net/bestia/messages/MessageTypeIdResolver.kt
package net.bestia.messages

import com.fasterxml.jackson.annotation.JsonTypeInfo.Id
import com.fasterxml.jackson.databind.DatabindContext
import com.fasterxml.jackson.databind.JavaType
import com.fasterxml.jackson.databind.jsontype.impl.TypeIdResolverBase
import com.fasterxml.jackson.databind.type.TypeFactory
import mu.KotlinLogging
import org.reflections.Reflections
import java.lang.IllegalArgumentException
import java.lang.reflect.Modifier
import java.util.*

private val LOG = KotlinLogging.logger { }

/**
 * Custom TypeId Resolver for message objects. Upon start it looks for all
 * messages which inherit from Message and determine their id. It stores it and
 * uses this serialize the messages.
 *
 * @author Thomas Felix
 */
class MessageTypeIdResolver : TypeIdResolverBase() {

  private val typeFactory = TypeFactory.defaultInstance()
  private lateinit var baseType: JavaType

  /**
   * Finds all IDs of the messages and registers them for later
   * identification.
   */
  override fun init(bt: JavaType) {
    super.init(bt)
    baseType = bt
  }

  override fun getMechanism(): Id {
    return Id.CUSTOM
  }

  override fun idFromValue(value: Any): String {
    return idFromValueAndType(value, value.javaClass)
  }

  override fun idFromValueAndType(value: Any, suggestedType: Class<*>): String {
    return classToId[suggestedType]
        ?: throw IllegalArgumentException("Could not get ID from class: $suggestedType")
  }

  override fun typeFromId(context: DatabindContext?, id: String?): JavaType {
    val clazz = idToClass[id]
    return typeFactory.constructSpecializedType(baseType, clazz)
  }

  companion object {
    private val idToClass = HashMap<String, Class<out MessageId>>()
    private val classToId = HashMap<Class<out MessageId>, String>()

    init {
      val reflections = Reflections("net.bestia.messages")
      val messages = reflections.getSubTypesOf(MessageId::class.java)

      messages.filter { !Modifier.isAbstract(it.modifiers) }
          .forEach {
            try {
              val messageId = it.getDeclaredField("MESSAGE_ID")
              val key = messageId.get(null) as String

              idToClass[key] = it
              classToId[it] = key

              LOG.trace { "Found Message.class: $key - $it" }
            } catch (e: Exception) {
              LOG.error(e) {
                "Could not get static MESSAGE_ID field from Message class: ${it.name}. " +
                    "Serialization and deserialization will fail."
              }
              System.exit(1)
            }
          }
    }
  }
}
